libcaf_net/src/net/web_socket/framing.cpp

```diff
@@ -250,4 +240,37 @@ void framing::ship_frame(std::vector<T>& buf) {
   buf.clear();
 }
 
+void framing::ship_closing_message(status code, std::string_view msg) {
+  auto code_val = static_cast<uint16_t>(code);
+  uint32_t mask_key = 0;
+  byte_buffer payload;
+  payload.reserve(msg.size() + 2);
+  payload.push_back(static_cast<std::byte>((code_val & 0xFF00) >> 8));
+  payload.push_back(static_cast<std::byte>(code_val & 0x00FF));
+  for (auto c : msg)
+    payload.push_back(static_cast<std::byte>(c));
+  if (mask_outgoing_frames) {
+    mask_key = static_cast<uint32_t>(rng_());
+    detail::rfc6455::mask_data(mask_key, payload);
+  }
+  down_->begin_output();
+  detail::rfc6455::assemble_frame(detail::rfc6455::connection_close, mask_key,
+                                  payload, down_->output_buffer());
+  down_->end_output();
+}
+
+void framing::abort_and_close_connection(sec reason, std::string_view msg) {
+  if (msg.empty())
+    up_->abort(make_error(reason));
+  else
+    up_->abort(make_error(reason, std::string(msg)));
```

Comment:
This duplicates logic from `shutdown`, which I would like to avoid. Did you consider implementing it like this?

```
template <class... Ts>
void abort_and_shutdown(sec reason, Ts&&... xs) {
  auto err = make_error(reason, std::forward<Ts>(xs)...);
  up_->abort(err);
  shutdown(err);
```

Naming it `abort_and_shutdown` would also keep the terminology consistent.

Reply:
Files a PR for the missing `string_view` support.